### ISSUE TITLE
skip_prefix => true doesn't remove ':' prefix separator

### ISSUE THREAD

**Issue by anton-yordanov** [joshuabowers/search_magic#12]
search_on :name, :skip_prefix => true
and for searchable_values we have  :  searchable_values: [":test", ":test1"]

**Comment by joshuabowers** [comment]
Are you using `:skip_prefix` on a field or on an association? While not enforced anywhere, the intention was that it was applicable only on associations; this would yield search keys which were not as cryptically arcane as they might otherwise be.

Using the option on a field would essentially nullify the search key for whatever value is stored in that field. Which, as you point out, yields entries in the `:searchable_values` array which are peculiar. Currently, this is not exactly supported: all steps except the `:search_for` work as intended; however, `:search_for` currently assumes that all values within `:searchable_values` have a non-empty key-prefix. While this could be conceivably changed to support what you are attempting, I would need to verify that it does not break any logic elsewhere or have other unintended consequences.

Note that such a modification would have at least one major impact on your models: you would essentially be throwing away the ability to target your search on that particular field.

**Comment by anton-yordanov** [comment]
I'm using it on association has_and_belongs_to_many. 
I use this option because search target  field is not important in my case 'autocomplete functionality'.  

Here is my model 

```
    class Asset
          include Mongoid::Document
          include Mongoid::Timestamps
          include SearchMagic

          field :name, type: String
          field :description, type: String

          has_and_belongs_to_many :keywords,   :index => true,  :dependent => :nullify
          has_and_belongs_to_many :categories, :index => true,  :dependent => :nullify

          search_on :name,        :skip_prefix => true
          search_on :keywords,    :skip_prefix => true
          search_on :categories,  :skip_prefix => true
    end
```

I'm not sure that  this is bug, or just I miss something but seems logical for me in :searchable_values array to have only text values when I use this :skip_prefix.
However i just write an a additional model that keep cache of all :searchable_values and updates via observer  that is a better solution in my case, because at first I just tried the easy way.

And at last, congratulations for this good gem.
Before I start to use it I try different gems that implement full text search for mongoid but yours is my favourite.  
I have only two wishes, configuration for word separator and 2-nd possibility to provide a file with black listed words

**Comment by joshuabowers** [comment]
I've added a feature-request issue for the [configurable word separator](joshuabowers/search_magic#13). If you would be so kind as to provide another issue with more information regarding your second request, I would be more than willing to take it into consideration for future development.

Your use-case might very well be considered a bug, at least depending upon how one defines the term. As the gem currently stands you essentially cannot search on data you requested to be searchable. As I mentioned in my first reply, a fix could certainly be possible.

How are you using the gem in a broader context? You mentioned 'autocomplete functionality'; does this mean you have an AJAX-based form for searching which autocompletes what users type into it? Or am I misunderstanding?

Out of curiosity, are you experiencing any unintuitive behavior with the handling of your associations? The current way the gem works would be to cause any fields from your `Keyword` and `Category` classes which have been marked as searchable to be subsumed into your `Asset` class' search space. As you are using `:skip_prefix`, those fields will be named as they are in the two associated classes.

**Comment by anton-yordanov** [comment]
At this moment I don't use :skip_prefix, because now I keep aggregated searchable_values by model types in separated model with removed target field to make  AJAX autocompleate.  At this way AJAX autocomplete is much responsive and the same time I can benefit from targeting my search on particular field in cases where I need this.

That is the result for searchable_values with :skip prefix, where asset have assigned  keywords
[":test", ":stamp", "name:keyword", "name:0", "name:keyword", "name:4"]
where the 1-st and 2-nd is    search_on :name, :skip_prefix => true
and the rest are    search_on :keywords, :skip_prefix => true
My expectations was to have just  ["test", "stamp", "keyword", "0", "keyword", "4"], at that moment I realized that there is no way to update properly searchable_values if you don't have key-prefix. I think that :skip_prefix must be used only once in per model

About the 2nd request,  I think that it will be useful to have some list with words which are not searchable. It can be  yaml file or just configuration array.

**Comment by joshuabowers** [comment]
There is now a feature request for a [black list](https://github.com/joshuabowers/search_magic/issues/16); it's currently planned for the 1.0.0 milestone, which should hopefully happen before too long.

I think there might be some misunderstanding off the overall role of **searchable_values** in the broader scope of SearchMagic: it never was intended to something that is looked at (nor updated!) by anything besides the gem itself. Values have prefixes within it for the reason of search: if you want to do a full text search, such is possible with prefixes. If you want to target your search to a specific field of your document, such is possible with prefixes. But altering the contents of the array so that prefixes are optional? That sort of defeats the purpose of being able to do a targeted search.

That said, the latest version of the gem, [0.2.0](http://rubygems.org/gems/search_magic/versions/0.2.0), does take a few steps to hopefully address a few of the issues you were running into. First, fields which have **skip_prefix** set will no longer have a prefix in **searchable_values**. Second, while this results in the loss of targeted search, you should be able to perform a full text search for those values.

About one of the other points you made in your last post: **skip_prefix** was never intended to be used the way you are using it. It was intended to be used only on associations which were searched upon, and, even then, only to simplify and clarify the prefixes. Depending upon how one codes, the prefixes are supposed to add information to the system, but this is not always the case. Removing (portions) of them is sometimes warranted to clean things up. However, it is important to understand which prefix is being discussed, here: namely, the prefix in question is the one coming from the association which is searched upon. 

It makes no sense to me to turn off prefix acquisition for the class which is defining the search. On the one hand, this is (relatively) meaningless: with the exception of the normal prefixing rules (which, I admit, might be debatable), a given class does not add any prefix to any of the fields it is asked to search on. Prefixes are added to the associations, or, more specifically, to fields that are made searchable in another document's context through an association.
